### PR TITLE
feat(research): H1 FLB data feasibility analysis

### PR DESCRIPTION
**Pull request by stone16**
## Summary

- Add standalone script (`scripts/flb_data_feasibility.py`) that fetches resolved Polymarket markets from the Gamma API and measures Favorite-Longshot Bias by probability decile with Wilson score 95% confidence intervals
- Sample gate: ≥100 resolved contracts in both <10% (longshot) and >90% (favorite) buckets
- Side semantics: longshot overpriced → BUY NO, favorite underpriced → BUY YES
- 31 tests covering decile assignment, Wilson interval math, FLB pattern detection, sample gate logic, and report generation

## Live Gamma API Run Evidence

**Total resolved markets:** 13

### Sample Gate Result

| Bucket | Count | Required | Status |
|--------|-------|----------|--------|
| Longshot (<10%) | 12 | ≥100 | ❌ |
| Favorite (>90%) | 1 | ≥100 | ❌ |

**Verdict: Gamma sample insufficient.** The Gamma API `closed=true` endpoint returns a small window of recently resolved markets. Dune Analytics on-chain data or a historical warehouse is needed for robust FLB measurement.

### FLB by Decile (13 markets)

| Decile | Range | N | Implied P | Actual Rate | FLB Gap | 95% CI |
|--------|-------|---|-----------|-------------|---------|--------|
| 0 | [0%-10%) | 12 | 0.3% | 0.0% | +0.3% | [0.0%, 24.3%] |
| 9 | [90%-100%] | 1 | 99.9% | 100.0% | -0.1% | [20.7%, 100.0%] |

Deciles 1-8: empty (no resolved markets in these price ranges from Gamma API).

### Categories

Sports: 8, Politics: 3, Other: 2. Total volume: $1.37M.

## Known Limitations (documented in script + report output)

1. **Entry price proxy:** Uses `lastTradePrice` (last trade before resolution), NOT a timestamped entry snapshot. For strategy P&L backtesting, we need price snapshots at a defined entry horizon (e.g., 24h before resolution).
2. **Data source:** Gamma API `closed=true` returns a small window. For ≥100 contracts per target bucket, Dune Analytics or historical warehouse is required.
3. **Feasibility only:** Bias-detection script, not a strategy backtest. Does not account for fees, slippage, or execution timing.

## Next Steps

- If Gamma sample insufficient (expected): create follow-up task for Dune/historical warehouse data source
- Once data is available: re-run script, verify FLB magnitude, proceed to full BacktestSpec + sweep pipeline

## Test plan

- [x] `uv run pytest tests/test_flb_data_feasibility.py -v` — 31/31 passed
- [x] `uv run pytest -q` — 946 passed (full suite, no regressions)
- [x] `uv run mypy scripts/flb_data_feasibility.py --strict` — clean

## Summary by CodeRabbit

* **Documentation**
  * Added FLB Data Feasibility Report with probability decile analysis, confidence intervals, and market category statistics for H1 evaluation.

* **New Features**
  * Added CLI tool to analyze resolved markets, compute FLB metrics per decile, generate feasibility reports, and export decile statistics to CSV.

* **Tests**
  * Added comprehensive test suite covering decile assignment, Wilson intervals, report generation, and edge case validation for feasibility analysis.

### CODE REVIEW

File: scripts/flb_data_feasibility.py [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+        if isinstance(prices_raw, str):
+            prices = json.loads(prices_raw)
+        else:
+            prices = prices_raw
+        if not isinstance(prices, list) or len(prices) < 2:
+            return None
+
+        yes_idx = outcomes.index("Yes")
+        no_idx = 1 - yes_idx  # binary market: the other index
+        yes_price = float(prices[yes_idx])
```
[the remark this answers is not in the document]
Comment (stone16) [reply scripts/flb_data_feasibility.py]: **Fixed** in ef1f9f1. Parser now requires exactly `{"Yes", "No"}` outcomes (len==2) before deriving `no_idx`. 3+ outcome rows are rejected with `return None`.

6 regression tests added in `TestBinaryOnlyParser`:
- `test_binary_yes_no_accepted` / `test_binary_no_yes_accepted` — valid binaries pass
- `test_three_outcomes_rejected` / `test_four_outcomes_rejected` / `test_yes_with_extra_outcome_rejected` / `test_single_outcome_rejected` — all non-binary variants rejected

Prices validation also tightened to `len(prices) != 2`.

File: scripts/flb_data_feasibility.py [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+
+    # Volume statistics.
+    volumes = [m.volume for m in markets]
+    if volumes:
+        lines.append("## Volume Statistics")
+        lines.append("")
+        lines.append(f"- **Min volume:** ${min(volumes):,.0f}")
+        lines.append(f"- **Median volume:** ${sorted(volumes)[len(volumes) // 2]:,.0f}")
+        lines.append(f"- **Max volume:** ${max(volumes):,.0f}")
+        lines.append(f"- **Total volume:** ${sum(volumes):,.0f}")
```
[the remark this answers is not in the document]
Comment (stone16) [reply scripts/flb_data_feasibility.py:590]: **Fixed** in ef1f9f1. Replaced `sorted(volumes)[len(volumes) // 2]` with `statistics.median(volumes)` for correct even-sample behavior.

2 regression tests added in `TestMedianVolume`:
- `test_even_count_averages_middle_two` — [100,200,300,400] → 250
- `test_odd_count_takes_middle` — [100,200,300] → 200

File: scripts/flb_data_feasibility.py [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+
+        # A resolved market has one side at 1.0 and the other at 0.0.
+        # Use a tolerance for floating-point representation.
+        if abs(yes_price - 1.0) < 0.01:
+            resolved_yes = True
+        elif abs(no_price - 1.0) < 0.01:
+            resolved_yes = False
+        else:
+            # Not clearly resolved — skip.
+            return None
```
[the remark this answers is not in the document]
Comment (stone16) [reply scripts/flb_data_feasibility.py:220]: Acknowledged. 50/50 resolution handling is out of scope for this patch (binary-only parser + median fix), but valid — will track as follow-up. Current parser correctly rejects these markets via the `prices[0] + prices[1]` proximity-to-1.0 check, which is conservative but safe for FLB measurement (FLB edge lives in extreme buckets, not 50/50). A future pass can add explicit 50/50 detection with configurable handling.

File: scripts/flb_data_feasibility.py [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+        no_idx = 1 - yes_idx  # binary market: the other index
+        yes_price = float(prices[yes_idx])
+        no_price = float(prices[no_idx])
+
+        # A resolved market has one side at 1.0 and the other at 0.0.
+        # Use a tolerance for floating-point representation.
+        if abs(yes_price - 1.0) < 0.01:
+            resolved_yes = True
+        elif abs(no_price - 1.0) < 0.01:
+            resolved_yes = False
```
[the remark this answers is not in the document]
Comment (stone16) [reply scripts/flb_data_feasibility.py]: **Fixed** in 06f8189.

Gamma API has no explicit settlement/resolution flag (checked: `resolved`, `finalized`, `settled`, `redeemed` — all `None`). The CLOB API tokens have `winner: true/false` but require per-market lookups via `clobTokenIds`, which is impractical for the bulk Gamma fetch path.

**Approach:** Tighten the price-based heuristic from 0.01 to 0.001 tolerance. This:
- ✅ Accepts AMM residual prices (0.999999/0.000001) — legitimate oracle-settled markets
- ✅ Accepts exact 0.0/1.0 payout vectors
- ❌ Rejects 0.995/0.005 near-resolution prices (the reported bug)
- ❌ Rejects 0.99/0.01 and mid-range prices
- ❌ Rejects degenerate 0.0/0.0 (cleared data from old markets like the 2020 Biden market)

**7 regression tests in `TestSettlementCriteria`:**
- `test_exact_payout_accepted` / `test_exact_reverse_payout_accepted` — 1.0/0.0 vectors
- `test_amm_residual_accepted` — 0.999999/0.000001
- `test_near_resolution_rejected` — **0.995/0.005 (key regression)**
- `test_99_1_percent_rejected` — 0.99/0.01
- `test_midrange_rejected` — 0.58/0.42
- `test_degenerate_zero_zero_rejected` — 0.0/0.0

**Documented limitation** in code comment: price-based heuristic is the best available proxy without an explicit oracle settlement indicator.
